### ISSUE TITLE
FP-optimization-of-pico-scale-generative-model-streaming-sample-expansion-0002

### ISSUE THREAD

**Issue by david-thrower** [david-thrower/cerebros-core-algorithm-alpha#220]
# From branch: #214 

# The Problem:

Each few samples when expanded for auto-regressive training create thousands of sample, label pairs. With the labels being (THOUSANDS_OF_EXPANDED_SAMPLES, 128_255)   ... the number of expanded samples X the tokenizer vocabulary, ... this is a few GB of expanded samples. This means that we need to stream the sample expansion at train time, so these can be expanded in batches of 2. That makes training infinitely scalable, in terms of training samples without additional RAM.  Alternatively we could use a sparse tensor.   

Try to integrate this as a first pass solution:

```
from typing import List, Iterator, Tuple

class StreamingInputIdsGenerator:
    """
    Straw man for x_train_tf that behaves exactly like tf.constant([[int, int, ... int], [int, int, ... int]])
    but generates data on-demand. When iterated over, it expands text samples in batches and returns
    one sequence at a time as tf.constant([int, int, ... int]).
    """
    
    def __init__(self, text_samples: List[str], 
                 text_expansion_batch_size: int = 2,
                 max_seq_length: int = 1536):
        """
        Initialize the streaming generator.
        
        Args:
            text_samples: List of text samples to process
            text_expansion_batch_size: Number of text samples to expand at once
            max_seq_length: Maximum sequence length for padding/truncation
        """
        self.text_samples = text_samples.copy()
        self.text_expansion_batch_size = text_expansion_batch_size
        self.max_seq_length = max_seq_length
        
        # Iterator state
        self.text_sample_indices = list(range(len(text_samples)))
        self.current_text_idx = 0
        self.rng = np.random.default_rng(42)
        self.rng.shuffle(self.text_sample_indices)
        
        # Buffer for expanded sequences
        self.sequence_buffer = []
        
        # Stats
        self.sequences_returned = 0
        
        # Pre-compute shape for compatibility
        self._shape = (len(text_samples) * 600, max_seq_length)  # Approximate
        
    def _expand_next_text_batch(self) -> int:
        """Expand the next batch of text samples and add to buffer."""
        if self.current_text_idx >= len(self.text_sample_indices):
            # Reset iterator if we've processed all samples
            self.current_text_idx = 0
            self.rng.shuffle(self.text_sample_indices)
        
        # Get next batch of text samples
        start_idx = self.current_text_idx
        end_idx = min(start_idx + self.text_expansion_batch_size, len(self.text_sample_indices))
        batch_indices = self.text_sample_indices[start_idx:end_idx]
        batch_samples = [self.text_samples[i] for i in batch_indices]
        
        print(f"Expanding {len(batch_samples)} text samples for input_ids")
        
        # Expand samples into token sequences (only input_ids)
        batch_input_ids, _, _ = prepare_data(batch_samples)
        
        # Add to buffer
        self.sequence_buffer.extend(batch_input_ids)
        
        sequences_generated = len(batch_input_ids)
        self.current_text_idx = end_idx
        
        print(f"  Generated {sequences_generated} input sequences")
        
        # Cleanup
        del batch_samples, batch_input_ids
        gc.collect()
        
        return sequences_generated
    
    def __iter__(self) -> Iterator[tf.Tensor]:
        """Make this iterable."""
        return self
    
    def __next__(self) -> tf.Tensor:
        """Return next sequence as tf.constant([int, int, ... int])."""
        # Ensure we have sequences in buffer
        while len(self.sequence_buffer) == 0:
            sequences_added = self._expand_next_text_batch()
            if sequences_added == 0:
                raise StopIteration
        
        # Take one sequence from buffer
        sequence = self.sequence_buffer.pop(0)
        self.sequences_returned += 1
        
        # Convert to TensorFlow tensor
        tensor_sequence = tf.constant(sequence, dtype=tf.int32)
        
        # Periodic cleanup
        if self.sequences_returned % 100 == 0:
            gc.collect()
            print(f"Returned {self.sequences_returned} sequences")
        
        return tensor_sequence
    
    def __getitem__(self, index: int) -> tf.Tensor:
        """
        Support indexing like a tensor - used by Cerebros.
        This is a simplified implementation for compatibility.
        """
        # For compatibility, we'll generate sequences on-demand
        # This is a simplified approach - in practice, you might want to cache
        if index >= self._shape[0]:
            raise IndexError("Index out of range")
        
        # Generate sequences until we reach the requested index
        while len(self.sequence_buffer) <= index:
            self._expand_next_text_batch()
        
        # Return the sequence at index
        sequence = self.sequence_buffer[index]
        return tf.constant(sequence, dtype=tf.int32)
    
    def __len__(self) -> int:
        """Return approximate length."""
        return self._shape[0]
    
    @property
    def shape(self) -> Tuple[int, int]:
        """Mimic tensor shape property."""
        return self._shape
    
    @property
    def dtype(self) -> tf.DType:
        """Mimic tensor dtype property."""
        return tf.int32

class StreamingLabelsGenerator:
    """
    Straw man for y_train_tf that behaves exactly like tf.constant([[float, float, ... float], [float, float, ... float]])
    but generates data on-demand and stays synchronized with StreamingInputIdsGenerator.
    """
    
    def __init__(self, text_samples: List[str], 
                 text_expansion_batch_size: int = 2,
                 vocabulary_size: int = VOCABULARY_SIZE):
        """
        Initialize the streaming labels generator.
        
        Args:
            text_samples: List of text samples to process (same as input generator)
            text_expansion_batch_size: Number of text samples to expand at once
            vocabulary_size: Size of vocabulary for one-hot encoding
        """
        self.text_samples = text_samples.copy()
        self.text_expansion_batch_size = text_expansion_batch_size
        self.vocabulary_size = vocabulary_size
        
        # Iterator state (must be synchronized with input generator)
        self.text_sample_indices = list(range(len(text_samples)))
        self.current_text_idx = 0
        self.rng = np.random.default_rng(42)
        self.rng.shuffle(self.text_sample_indices)
        
        # Buffer for expanded labels
        self.label_buffer = []
        
        # Stats
        self.labels_returned = 0
        
        # Pre-compute shape for compatibility
        self._shape = (len(text_samples) * 600, vocabulary_size)  # Approximate
        
    def _expand_next_text_batch(self) -> int:
        """Expand the next batch of text samples and add labels to buffer."""
        if self.current_text_idx >= len(self.text_sample_indices):
            # Reset iterator if we've processed all samples
            self.current_text_idx = 0
            self.rng.shuffle(self.text_sample_indices)
        
        # Get next batch of text samples
        start_idx = self.current_text_idx
        end_idx = min(start_idx + self.text_expansion_batch_size, len(self.text_sample_indices))
        batch_indices = self.text_sample_indices[start_idx:end_idx]
        batch_samples = [self.text_samples[i] for i in batch_indices]
        
        print(f"Expanding {len(batch_samples)} text samples for labels")
        
        # Expand samples into token sequences (only labels)
        _, batch_labels, _ = prepare_data(batch_samples)
        
        # Add to buffer
        self.label_buffer.extend(batch_labels)
        
        labels_generated = len(batch_labels)
        self.current_text_idx = end_idx
        
        print(f"  Generated {labels_generated} label sequences")
        
        # Cleanup
        del batch_samples, batch_labels
        gc.collect()
        
        return labels_generated
    
    def __iter__(self) -> Iterator[tf.Tensor]:
        """Make this iterable."""
        return self
    
    def __next__(self) -> tf.Tensor:
        """Return next label as tf.constant([float, float, ... float])."""
        # Ensure we have labels in buffer
        while len(self.label_buffer) == 0:
            labels_added = self._expand_next_text_batch()
            if labels_added == 0:
                raise StopIteration
        
        # Take one label from buffer
        label = self.label_buffer.pop(0)
        self.labels_returned += 1
        
        # Convert to TensorFlow tensor
        tensor_label = tf.constant(label, dtype=tf.float32)
        
        # Periodic cleanup
        if self.labels_returned % 100 == 0:
            gc.collect()
            print(f"Returned {self.labels_returned} labels")
        
        return tensor_label
    
    def __getitem__(self, index: int) -> tf.Tensor:
        """
        Support indexing like a tensor - used by Cerebros.
        """
        if index >= self._shape[0]:
            raise IndexError("Index out of range")
        
        # Generate labels until we reach the requested index
        while len(self.label_buffer) <= index:
            self._expand_next_text_batch()
        
        # Return the label at index
        label = self.label_buffer[index]
        return tf.constant(label, dtype=tf.float32)
    
    def __len__(self) -> int:
        """Return approximate length."""
        return self._shape[0]
    
    @property
    def shape(self) -> Tuple[int, int]:
        """Mimic tensor shape property."""
        return self._shape
    
    @property
    def dtype(self) -> tf.DType:
        """Mimic tensor dtype property."""
        return tf.float32

# Factory function to create the streaming data objects
def create_streaming_training_data(text_samples: List[str], 
                                 text_expansion_batch_size: int = 2) -> Tuple[List, List]:
    """
    Create streaming data objects that can be passed directly to Cerebros.
    
    Args:
        text_samples: List of text samples to process
        text_expansion_batch_size: Number of text samples to expand at once (2 for ~1GB memory usage)
        
    Returns:
        Tuple of (x_train_packaged, y_train_packaged) ready for Cerebros
    """
    print(f"Setting up streaming data for {len(text_samples)} text samples")
    print(f"Text expansion batch size: {text_expansion_batch_size}")
    
    # Create streaming generators
    x_train_streaming = StreamingInputIdsGenerator(
        text_samples=text_samples,
        text_expansion_batch_size=text_expansion_batch_size,
        max_seq_length=MAX_SEQ_LENGTH
    )
    
    y_train_streaming = StreamingLabelsGenerator(
        text_samples=text_samples,
        text_expansion_batch_size=text_expansion_batch_size,
        vocabulary_size=VOCABULARY_SIZE
    )
    
    # Package as lists for Cerebros compatibility
    x_train_packaged = [x_train_streaming]
    y_train_packaged = [y_train_streaming]
    
    return x_train_packaged, y_train_packaged

```

## So I don't lose it if I need to refine it, the endless prompt needed to get this written ...

```
Here is what I really need. I need 2 classes:

1. A straw man for x_train_tf that was defined as tf.constant(X_train_final, tf.int32) (tf.constant([[int, int, ... int], [int, int, ... int]])) of shape (TOTAL_NUMBER_OF_SAMPLES_AFTER_EXPANSION, MAX_SEQ_LEN) ... that when you iterate over it, it will expand the samples (list[str], configured instantiation time) in a batch of say 2 using package_non_instruct_text() and assign the result to the buffer . It will return 1 of the items off the buffer each def __next__(self): tf.constant([float, float, float]) .., or whatever is needed so this this can be packaged as:

x_train_packaged = [x_train_tf] and passed directly to CerebrosSimpleRandomSearch without a wrapper. There are structural reasons why the wrapper won't be practical. It creates a model, calls model.fit() on the data, gets the metrics, saves the model to disk, the garbage collects the model and assigns the next model it creates to the same "self variable". Once it has finished that round of training, the same model  will not be available to continue training from outside (unless we load the model from disk). If we did, it undermines what the NAS is doing (automating finding the best model on a small data set in Phase I training, which will later be loaded from disk for phase II training). The problem to solve now is to complete the phase I training on a small data set (synthetic text book, open source literature, ...) to proove concept and select which model is a suitable foundation (lowest perplexity, best BLEU, ROUGE) to go to phase II and Phase III training on synthetic social communication, synthetic business and professional documents, and instruct / reasoning fine tuning and RLHF... The reasoning that you have recommended thus far are right for the phase II and III training once we have only 1 model to train directly, but in the phase I stage, we need to make a generator that acts as a straw man for the exact data structure used so far to get there.       

As for the labels, we need a straw man for  y_train_tf defined as tf.constant(y_train_final, tf.float32) # Shape of (TOTAL_NUMBER_OF_SAMPLES_AFTER_EXPANSION, VOCABULARY_SIZE) that can be directly packaged as

y_train_packaged = [y_train_tf]

# We also need to keep these synchronized where the label and sample are in lock step (sample[i] is paired with label[i] such that label[i] represents the logit that would be correct for the NEXT word / non-pad-token, unless the next token is the first occurrence of a pad token, in which case it is the logit for the pad_token and the sub - samples for this test sample are complete and it is time  the next sub-sample will be the first sample from the next text sample ....)
 
Here is the original code for reference: ... 

## A copy of the script goes here ....

...

```

**Comment by david-thrower** [comment]
Close to #230
